### PR TITLE
executor: add the step handler

### PR DESCRIPTION
**Pull request by cmatsuoka**
The step handler takes care of the execution of a step, using either
a built-in set of actions to be taken, or executing a user-defined
script defined in the part specification. User-defined scripts may also
call the built-in handler for a step by invoking a control utility.

This class implements the built-in handlers and a FIFO-based mechanism
and API to be used by the external control utility to communicate with
the running instance.

The control API server is a direct port of the implementation used in
Snapcraft, and should be refactored later.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

File: craft_parts/executor/step_handler.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+from craft_parts.parts import Part
+from craft_parts.plugins import Plugin
+from craft_parts.sources import SourceHandler
+from craft_parts.steps import Step
+from craft_parts.utils import file_utils
+
+from . import environment, filesets
+from .filesets import Fileset
+
+FilesAndDirs = namedtuple("FilesAndDirs", ["files", "dirs"])
```
Comment (cjp256) [inline craft_parts/executor/step_handler.py:44]:
```suggestion
@dataclasses.dataclass
class FilesAndDirs:
  files: Set[str]
  dirs: Set[str]
```

**Review by cjp256 — APPROVED** [review]

File: craft_parts/executor/step_handler.py (hunk 1 above)
Comment (cmatsuoka) [reply craft_parts/executor/step_handler.py:44]: In this case we're simply organizing the tuple returned by handler functions, and adding the overhead of a full dataclass is probably not worth it.

Comment (cjp256) [reply craft_parts/executor/step_handler.py:44]: > In this case we're simply organizing the tuple returned by handler functions, and adding the overhead of a full dataclass is probably not worth it.

The advantage to the dataclass over namedtuple is type information.  Which could be helpful in development.  But not really critical.

File: craft_parts/executor/step_handler.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+) -> None:
+    """Rewrite the prefix entry in a pkg-config file.
+
+    :param root: The root to add to the configuration prefix.
+    :param pkg_config_file: The pkg-config file to process.
+    :param prefix_trim: The initial path to remove from the configuration prefix.
+    """
+    pattern_trim = None
+    if prefix_trim:
+        pattern_trim = re.compile("^prefix={}(?P<prefix>.*)".format(prefix_trim))
```
Comment (sergiusens) [inline craft_parts/executor/step_handler.py:347]: maybe take the opportunity to fix https://bugs.launchpad.net/snapcraft/+bug/1916281 :-)

**Review by sergiusens — APPROVED** [review]
Looks good, not a fan of returning sets for pull and build unless we have a reason I am not aware of in the future (or maybe it is just for symmetry)

**Comment by cmatsuoka** [comment]
> Looks good, not a fan of returning sets for pull and build unless we have a reason I am not aware of in the future (or maybe it is just for symmetry)

It was done this way to allow uniform call protocol for all steps, but it can be refactored later if needed.

File: craft_parts/executor/step_handler.py (hunk 2 above)
Comment (cmatsuoka) [reply craft_parts/executor/step_handler.py:347]: Will add a note to the code and fix in a different PR.

**Pull request merged**